### PR TITLE
Implement gen_enforced_dependency_range/3

### PR DESCRIPTION
**Pull request by bgotink**
Implements #68

### CODE REVIEW

File: packages/plugin-constraints/sources/Constraints.ts [hunk 1]
```diff
@@ -132,6 +122,9 @@ export class Constraints {
     // (Cwd, DependencyIdent, DependencyType, Reason)
     declarations += `gen_invalid_dependency(_, _, _, _) :- false.\n`;
 
+    // (Cwd, Path, Value)
+    declarations += `gen_workspace_field_requirement(_, _, _) :- false.\n`;
```
Comment (arcanis) [inline packages/plugin-constraints/sources/Constraints.ts]: Thinking about it it's too bad the names are so different ... maybe we should uniformize them? ie one of those:

- `gen_enforced_dependency` / `gen_enforced_field`
- `gen_workspace_dependency_requirement` / `gen_workspace_field_requirement`

**Comment by arcanis** [comment]
Nice! It's also missing the autofix part, right?

File: packages/plugin-constraints/sources/Constraints.ts (hunk 1 above)
Comment (bgotink) [reply packages/plugin-constraints/sources/Constraints.ts]: Yeah it irked me as well, I'll change it to `gen_enforced_dependency`/`gen_enforced_field`

**Comment by bgotink** [comment]
The `constraint fix` command hasn't been updated yet indeed. I'd like to keep that out of scope for now, because that needs some more thought. For example, with rule

```pl
gen_enforced_field(_, sideEffects, false).

gen_enforced_field(_, 'publishConfig.registry', null).
```

- do we want the boolean `false` or the string `'false'` for the `sideEffects` property?
- if we delete `publishConfig.registry` and `publishConfig` is now empty, do we want to remove the entire `publishConfig` property or keep it empty?

**Comment by bgotink** [comment]
I'm also struggling with a rule I'd like to use: "this property must be present, though I don't care about its value" (actually, it has to be an object, but lets leave that out of the discussion for now)

I can't do that without a `gen_invalid_field`-like rule. Unlike dependencies (#122) I see a real need for the "invalid" rule here.

**Comment by arcanis** [comment]
Fine by me to keep `fix` out of scope for now. For the record:

> do we want the boolean `false` or the string `'false'` for the `sideEffects` property?

I think an acceptable solution might be to send it through `JSON.parse` and fallback to string if it doesn't parse. This way if users explicitly want a string to be sent they can use `'"false"'`.

> if we delete `publishConfig.registry` and `publishConfig` is now empty, do we want to remove the entire `publishConfig` property or keep it empty?

The `Manifest#exportTo` function already removes the fields once they are unused; I think it's fine if we keep this trend 🤔

> I'm also struggling with a rule I'd like to use: "this property must be present, though I don't care about its value" (actually, it has to be an object, but lets leave that out of the discussion for now)

Do you think it would be possible to represent that through a special term? For example something like this:

```
gen_enforced_field(Cwd, 'publishConfig', any) :-
  workspace(Cwd).
```

Where the fix engine would know that `any` cannot be automatically generated and would thus generate the same kind of unfixable errors as conflicts.

**Comment by arcanis** [comment]
> The `Manifest#exportTo` function already removes the fields once they are unused; I think it's fine if we keep this trend 🤔

Or if we support re-evaluation after each fix (which you mentioned in https://github.com/yarnpkg/berry/issues/122#issuecomment-489403980) it's just a matter of writing something like this:

```
gen_enforced_field(Cwd, Name, null) :-
  workspace_field(Cwd, Name, '{}').
```

**Pull request merged**